core/drivers/hisi_trng.c

```diff
@@ -0,0 +1,94 @@
+// SPDX-License-Identifier: GPL-2.0
+/* Copyright (c) 2021 HiSilicon Limited. */
+
+#include <initcall.h>
+#include <io.h>
+#include <malloc.h>
+#include <mm/core_memprot.h>
+#include <platform_config.h>
+#include <rng_support.h>
+#include <kernel/spinlock.h>
+
+#define HTRNG_RANDATA_REG 0xF0
+#define HTRNG_BYTES 4
+
+#define POLL_PERIOD 10
+#define POLL_TIMEOUT 1000
+
+struct hisi_trng {
+	vaddr_t base;
+};
+
+static unsigned int trng_lock = SPINLOCK_UNLOCK;
+static struct hisi_trng *trng_dev;
+
+static TEE_Result trng_read(uint32_t *val)
+{
+	uint32_t exceptions = 0;
+
+	exceptions = cpu_spin_lock_xsave(&trng_lock);
+	if (IO_READ32_POLL_TIMEOUT(trng_dev->base + HTRNG_RANDATA_REG,
+				   val, val, POLL_PERIOD, POLL_TIMEOUT)) {
+		EMSG("Hardware busy.");
+		cpu_spin_unlock_xrestore(&trng_lock, exceptions);
+		return TEE_ERROR_BUSY;
```

Comment:
How about using a local variable `TEE_Result res = TEE_SUCCESS` and assigning it with `TEE_ERROR_BUSY` here? That way you could remove the `cpu_spin_unlock_xrestore()` at the line above.